pilot/pkg/workloadentry/controller.go

```diff
@@ -0,0 +1,308 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/workloadentry/internal/autoregistration"
+	"istio.io/istio/pilot/pkg/workloadentry/internal/health"
+	workloadentrystore "istio.io/istio/pilot/pkg/workloadentry/internal/store"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/config/schema/gvk"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/queue"
+	istiolog "istio.io/pkg/log"
+)
+
+const (
+	// maxRetries is the number of times a service will be retried before it is dropped out of the queue.
+	// With the current rate-limiter in use (5ms*2^(maxRetries-1)) the following numbers represent the
+	// sequence of delays between successive queuings of a service.
+	//
+	// 5ms, 10ms, 20ms, 40ms, 80ms
+	maxRetries = 5
+)
+
+var log = istiolog.RegisterScope("wle", "wle controller debugging")
+
+type HealthEvent = health.HealthEvent
+
+var WorkloadControllerAnnotation = workloadentrystore.WorkloadControllerAnnotation
+
+type WorkloadEntryCleaner interface {
+	GetCleanupGracePeriod() time.Duration
+	ShouldCleanup(wle *config.Config, maxConnectionAge time.Duration) bool
+	Cleanup(wle *config.Config, periodic bool)
+}
+
+type Controller struct {
+	instanceID string
+	// TODO move WorkloadEntry related tasks into their own object and give InternalGen a reference.
+	// store should either be k8s (for running pilot) or in-memory (for tests). MCP and other config store implementations
+	// do not support writing. We only use it here for reading WorkloadEntry/WorkloadGroup.
+	store model.ConfigStoreController
+
+	// Note: unregister is to update the workload entry status: like setting `DisconnectedAtAnnotation`
+	// and make the workload entry enqueue `cleanupQueue`
+	// cleanup is to delete the workload entry
+
+	// queue contains workloadEntry that need to be unregistered
+	queue controllers.Queue
+	// cleanupLimit rate limit's auto registered WorkloadEntry cleanup calls to k8s
+	cleanupLimit *rate.Limiter
+	// cleanupQueue delays the cleanup of auto registered WorkloadEntries to allow for grace period
+	cleanupQueue queue.Delayed
+
+	mutex sync.Mutex
+	// record the current adsConnections number
+	// note: this is to handle reconnect to the same istiod, but in rare case the disconnect event is later than the connect event
+	// keyed by proxy network+ip
+	adsConnections map[string]uint8
+
+	// maxConnectionAge is a duration that workload entry should be cleaned up if it does not reconnects.
+	maxConnectionAge time.Duration
+
+	wleStore         *workloadentrystore.Controller
+	healthController *health.Controller
+	autoregistration *autoregistration.Controller
+}
+
+// NewController create a controller which manages workload lifecycle and health status.
+func NewController(store model.ConfigStoreController, instanceID string, maxConnAge time.Duration) *Controller {
+	if features.WorkloadEntryAutoRegistration || features.WorkloadEntryHealthChecks {
```

Comment:
Consider negating this condition to avoid having the whole function body within the conditional

Reply:
This PR will be replaced by https://github.com/istio/istio/pull/44834.

This is the [existing code](https://github.com/istio/istio/blob/619e8c3842b7fcf55bc6addf879e57fecab854e3/pilot/pkg/autoregistration/controller.go#L153). I'm planning to leave it "as is".